buildSrc/src/main/kotlin/Publish.kt

```diff
@@ -1,10 +1,107 @@
+
 import org.gradle.api.Project
+import org.gradle.api.publish.PublishingExtension
+import org.gradle.api.publish.maven.MavenPublication
+import org.gradle.api.tasks.bundling.Jar
 import org.gradle.kotlin.dsl.apply
+import org.gradle.kotlin.dsl.configure
+import org.gradle.kotlin.dsl.extra
+import org.gradle.kotlin.dsl.register
+import org.gradle.kotlin.dsl.the
+import org.gradle.plugins.signing.SigningExtension
 
 fun Project.applyMavenPublish() {
 
     apply(plugin = "maven-publish")
+    apply(plugin = "signing")
+
+    val versionName: String = property("VERSION_NAME") as String
 
     group = "ag.granular"
-    version  = property("VERSION") as String
-}
\ No newline at end of file
+    version = versionName
+
+    val isReleaseBuild = !versionName.contains("SNAPSHOT")
+
+    val emptySourcesJar = tasks.register<Jar>("emptySourcesJar") {
+        classifier = "sources"
+    }
+
+    val emptyJavadocJar = tasks.register<Jar>("emptyJavadocJar") {
+        classifier = "javadoc"
+    }
+
+    val releaseRepositoryUrl: String = findProperty("RELEASE_REPOSITORY_URL") as String?
+        ?: "https://oss.sonatype.org/service/local/staging/deploy/maven2/"
+
+    val snapshotRepositoryUrl: String =
+        findProperty("SNAPSHOT_REPOSITORY_URL") as String? ?: "https://oss.sonatype.org/content/repositories/snapshots/"
+
+    val repositoryUsername: String = System.getenv("SONATYPE_USERNAME") ?: ""
```

Comment:
It might be nice to place all environment variable keys at the top as `def`s, as convenience for folks who are configuring their machine for the first time.